### PR TITLE
Stop hard-coding ClearlyDefined providers

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: model/src/main/kotlin/PackageProvider.kt [hunk 1]
```diff
@@ -0,0 +1,147 @@
+/*
```
Comment (fviernau) [inline model/src/main/kotlin/PackageProvider.kt:1]: commit-msg: "This is will" -> "This will"

File: model/src/main/kotlin/PackageProvider.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ * artifacts under the same [Identifier]. So if in question, the provider must be given in addition to the [Identifier].
+ */
+enum class PackageProvider(
+    /**
+     * The regular expression patterns that match all URLs for a specific provider.
+     */
+    vararg urlPatterns: String,
+
+    /**
+     * The name ClearlyDefined uses for this provider. If a provider if not available in ClearlyDefined (yet), use an
```
Comment (fviernau) [inline model/src/main/kotlin/PackageProvider.kt:34]: "If a provider if" -> "If a provider i**s**"

File: model/src/main/kotlin/PackageProvider.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        clearlyDefinedName = "rubygems"
+    ),
+
+    // The following have no matching ClearlyDefined provider.
+    ANDROID(
+        "^https?://android\\.googlesource\\.com/"
+    ),
+    APACHE_GIT(
+        "^https?://gitbox\\.apache\\.org/repos/.+\\.git",
+        "^https?://git-wip-us\\.apache\\.org/repos/.+\\.git"
```
Comment (fviernau) [inline model/src/main/kotlin/PackageProvider.kt:104]: I fail to understand the point of `.+`. Could you help?

File: model/src/main/kotlin/PackageProvider.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    UNKNOWN(
+        ".*"
+    );
+
+    private val urlRegexes by lazy { urlPatterns.map { it.toRegex() } }
+
+    companion object {
+        fun get(url: String) = enumValues<PackageProvider>().find { provider ->
+            provider.urlRegexes.any { it.containsMatchIn(url) }
+        } ?: UNKNOWN
```
Comment (fviernau) [inline model/src/main/kotlin/PackageProvider.kt:141]: UNKOWN is guaranteed to match, so `?: UNKOWN` should never happen. Would it make sense to express this with `!!` or rather `checkNotNull()` ?

File: model/src/main/kotlin/PackageProvider.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    // The following matches any (unknown) provider.
+    UNKNOWN(
+        ".*"
+    );
+
+    private val urlRegexes by lazy { urlPatterns.map { it.toRegex() } }
+
+    companion object {
+        fun get(url: String) = enumValues<PackageProvider>().find { provider ->
```
Comment (fviernau) [inline model/src/main/kotlin/PackageProvider.kt:139]: I'm not 100% certain if the patterns match. (E.g. wondering if patterns should end with `.*`. Could you add a unit test to cover at least one or two cases, maybe picking one with one of the more complicated patterns?

File: model/src/main/kotlin/PackageProvider.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    ),
+
+    // The following matches any (unknown) provider.
+    UNKNOWN(
+        ".*"
+    );
+
+    private val urlRegexes by lazy { urlPatterns.map { it.toRegex() } }
+
+    companion object {
```
Comment (fviernau) [inline model/src/main/kotlin/PackageProvider.kt:138]: optional: Imo explicit return types would be nice to have nullability obvious.

File: model/src/main/kotlin/utils/Extensions.kt [hunk 7]
```diff
@@ -27,48 +27,67 @@ import org.ossreviewtoolkit.clients.clearlydefined.Provider
...
 import org.ossreviewtoolkit.utils.common.percentEncode
 
 internal fun TextLocation.prependPath(prefix: String): String =
     if (prefix.isBlank()) path else "${prefix.removeSuffix("/")}/$path"
 
 /**
- * Map an [Package] to a ClearlyDefined [ComponentType] and [Provider]. Note that an
- * [identifier's type][Identifier.type] in ORT currently implies a default provider. Return null if a mapping is not
- * possible.
+ * Map a [Package] [identifier's type][Identifier.type] to a ClearlyDefined [ComponentType]. Return null if a mapping is
```
Comment (fviernau) [inline model/src/main/kotlin/utils/Extensions.kt:41]: maybe instead: "Map [type][Identifier.type] of a [package identifier] [Package.id] to [..]"

File: model/src/main/kotlin/PackageProvider.kt [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model
```
Comment (fviernau) [inline model/src/main/kotlin/PackageProvider.kt:20]: Is this supposed to be used only from `ClearlyDefinedPackageCurationProvider`? If so, shouldn't it be put next to it (in the analyzer) ?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/PackageProvider.kt (hunk 3 above)
Comment (sschuberth) [reply model/src/main/kotlin/PackageProvider.kt:104]: I want to ensure that the URLs end with ".git", and that something comes between that and the base URL / base path.

File: model/src/main/kotlin/PackageProvider.kt (hunk 5 above)
Comment (sschuberth) [reply model/src/main/kotlin/PackageProvider.kt:139]: FYI, I created a temporary unit test when crafting these patterns, and that test went over all remote artifact URLs or VCS URLs in all of our committed ORT results files. So I'm pretty sure the patterns cover all relevant cases. But the test simply was not in shape to be committed.

File: model/src/main/kotlin/PackageProvider.kt (hunk 8 above)
Comment (sschuberth) [reply model/src/main/kotlin/PackageProvider.kt:20]: Strictly speaking #155 is about adding the "(package) provider" concept to ORT independently of ClearlyDefined. So while I agree the primary use-case to be in context with ClearlyDefined, I do not want to rule out other use-cases at this point by moving the class.

Comment (sschuberth) [reply model/src/main/kotlin/PackageProvider.kt:20]: But you're right that I should separate the ClearlyDefined-specific code more from the enum. Will do.

**Review by fviernau — DISMISSED** [review]

**Comment by sschuberth** [comment]
I had to rebase to resolve a trivial merge conflict in imports.

**Review by fviernau — APPROVED** [review]

**Comment by sschuberth** [comment]
Merging despite the unrelated test failure.

**Pull request merged**